### ISSUE TITLE
Issues connecting to nzbget 17 over TLS with Chrome

### ISSUE THREAD

**Issue by sselph** [nzbget/nzbget#243]
I installed nzbget on FreeBSD in a jail using the 17.0-testing-r1735 script in the releases and configured it to run with TLS. I was able to connect with Safari but not with Chrome. On Chrome I got server "unexpectedly closed the connection." I tried compiling the same version from source with debug enabled and saw this in the logs:

```
Sun Jul 17 23:23:18 2016    32527   34410098688 DEBUG   Creating Connection (Connection.cpp:129:Connection)
Sun Jul 17 23:23:18 2016    32527   34410098688 DEBUG   Creating Thread (Thread.cpp:83:Thread)
Sun Jul 17 23:23:18 2016    32527   34410098688 DEBUG   Starting Thread (Thread.cpp:93:Start)
Sun Jul 17 23:23:18 2016    32527   34410098688 DEBUG   Accepting connection (Connection.cpp:439:Accept)
Sun Jul 17 23:23:18 2016    32527   34423334912 DEBUG   Entering Thread-func (Thread.cpp:164:thread_handler)
Sun Jul 17 23:23:18 2016    32527   34423334912 DEBUG   Starting TLS (Connection.cpp:885:StartTls)
Sun Jul 17 23:23:18 2016    32527   34410098688 DEBUG   Creating Connection (Connection.cpp:129:Connection)
Sun Jul 17 23:23:18 2016    32527   34410098688 DEBUG   Creating Thread (Thread.cpp:83:Thread)
Sun Jul 17 23:23:18 2016    32527   34410098688 DEBUG   Starting Thread (Thread.cpp:93:Start)
Sun Jul 17 23:23:18 2016    32527   34410098688 DEBUG   Accepting connection (Connection.cpp:439:Accept)
Sun Jul 17 23:23:18 2016    32527   34423320576 DEBUG   Entering Thread-func (Thread.cpp:164:thread_handler)
Sun Jul 17 23:23:18 2016    32527   34423320576 DEBUG   Starting TLS (Connection.cpp:885:StartTls)
Sun Jul 17 23:23:18 2016    32527   34423334912 DEBUG   TLS handshake failed: error:00000000:lib(0):func(0):reason(0) (TlsSocket.cpp:233:ReportError)
Sun Jul 17 23:23:18 2016    32527   34423334912 DEBUG   Could not establish secure connection to web-client: Start TLS failed (RemoteServer.cpp:133:Run)
Sun Jul 17 23:23:18 2016    32527   34423334912 DEBUG   Thread-func exited (Thread.cpp:170:thread_handler)
Sun Jul 17 23:23:18 2016    32527   34423334912 DEBUG   Autodestroying Thread-object (Thread.cpp:176:thread_handler)
Sun Jul 17 23:23:18 2016    32527   34423334912 DEBUG   Disconnecting (Connection.cpp:188:Disconnect)
Sun Jul 17 23:23:18 2016    32527   34423334912 DEBUG   Do disconnecting (Connection.cpp:778:DoDisconnect)
Sun Jul 17 23:23:18 2016    32527   34423334912 DEBUG   Destroying Connection (Connection.cpp:147:~Connection)
Sun Jul 17 23:23:18 2016    32527   34423334912 DEBUG   Disconnecting (Connection.cpp:188:Disconnect)
Sun Jul 17 23:23:18 2016    32527   34423334912 DEBUG   Destroying Thread (Thread.cpp:88:~Thread)
Sun Jul 17 23:23:18 2016    32527   34423320576 DEBUG   TLS handshake failed: error:00000000:lib(0):func(0):reason(0) (TlsSocket.cpp:233:ReportError)
Sun Jul 17 23:23:18 2016    32527   34423320576 DEBUG   Could not establish secure connection to web-client: Start TLS failed (RemoteServer.cpp:133:Run)
```

I installed version 16.4 using precompiled binaries with the same TLS config and it is working.

**Comment by hugbug** [comment]
> I installed version 16.4 using precompiled binaries with the same TLS config and it is working.

Precompiled binaries from where? V16 didn't offer binaries for FreeBSD.

**Comment by sselph** [comment]
Sorry, the one from packages (`sudo pkg install nzbget`) I'm new to FreeBSD so my terminology might be off.

**Comment by hugbug** [comment]
Can you please tell me the dependencies of nzbget package which works: does it use OpenSSL or GnuTLS?

**Comment by sselph** [comment]
The 16.4 package is using GnuTLS. I tried both when compiling 17, I think I only enabled debug with openSSL.

**Comment by hugbug** [comment]
Does Chrome run on the same machine? Can this be reproduced with Chrome on Windows or Mac (when connecting to nzbget running on FreeBSD)?

**Comment by sselph** [comment]
This was from a different machine on the same LAN, running OS X and Chrome beta. I didn't try other OSs or Chrome versions but could try a linux and/or chromebook installation this evening as well as different versions or other browsers, etc. I did try connecting from an app in Android that wasn't able to connect over TLS as well but didn't log anything from it to see anything, it was still in the phase of "I must be doing something wrong." I can't test from FreeBSD since there is no GUI on this system and I don't own windows to be able to test that.

**Comment by sselph** [comment]
I apologize, something came up and I didn't have time to do more troubleshooting but should have time this evening.

**Comment by hugbug** [comment]
I've made a test on FreeBSD 10.2 (VM), connecting from Chrome running on OS X, both current stable and current beta. It worked.

So may be it is a configuration issue on your side? Something with jail or firewall. I don't have better ideas at the moment. The most strange thing is that it works for you in Safari at the same time.

**Comment by sselph** [comment]
Well I apologize for wasting your time. I had the same thought that it may have been my config but when it worked with Safari I thought there might be more too it. I guess it could still be related to my config and Safari is just way more forgiving than Chrome. I'll go ahead and close this and if I get more information that does point to something specific, I'll let you know.

**Comment by hugbug** [comment]
No need to apologize. The issue may exist; it's difficult for me to do something about it at the moment, without being able to reproduce and debug it.

**Comment by sselph** [comment]
I did more troubleshooting but didn't come up with much. I'm going to put this information here in case it comes up again or I decide to look further. My current conclusion is I've got something misconfigured with openssl on freebsd.

I tested with the v17 provided in the releases, I compared it with the v16.4 in pkg but since one is using openssl and the other gnutls the comparisons weren't very good. My observation is the TLS settings for gnutls was much better than openssl (A vs F on ssllabs) To give a better comparison I compared to v16.4 on my linux box that was also using openssl and everything looked similar. The only difference was there was an extra cipher suite one of them. I tried connecting to both openssl instances using `openssl s_client -connect` and diff'd the output. They were virtually identical(only time and key differed) but openssl complained with `140419733001880:error:140790E5:SSL routines:ssl23_write:ssl handshake failure:s23_lib.c:177:` in both cases.

As for browsers Chrome, Firefox and Safari connected to v16.4 using openssl on my linux machine. Safari but not Chrome and Firefox connected to v17 using openssl on my FreeBSD setup.
